qualtran/bloqs/arithmetic/Error propagation due to fixed point arithmetic.ipynb

```diff
@@ -0,0 +1,203 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "6d12f4fc-8129-4c12-a00f-fe52aa4888e5",
+   "metadata": {},
+   "source": [
+    "## Fixed point arithmetic error propagation analysis using fxpmath\n",
+    "\n",
+    "In this notebook, we use library fxpmath to analyze the propagation of errors due to fixed point arithmetic for algorithms described in [Compilation of Fault-Tolerant Quantum Heuristics for Combinatorial Optimization](https://arxiv.org/abs/2007.07391). \n",
+    "\n",
+    "Qualtran should provide tools to enable such analysis for algorithms that use fixed point arithmetic Bloqs as subroutines. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "id": "92cdfbaa-1e5f-4aac-9fdd-4d4015fd384a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from fxpmath import Fxp\n",
+    "import numpy as np\n",
+    "\n",
+    "# Initial setup - Helpers\n",
+    "def fxp(x: float, d: int) -> Fxp:\n",
+    "    dtype = f'fxp-u{d}/{d}'\n",
+    "    ret = Fxp(x, dtype=dtype, op_sizing='same', const_op_sizing='same', shifting='trunc')\n",
+    "    return ret\n",
+    "\n",
+    "def assert_allclose(x: Fxp, y: float, eps: float):\n",
+    "    np.testing.assert_allclose(x.get_val(), y, atol=eps)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "33cfaea3-f594-48f8-a846-4fa30bb9774b",
+   "metadata": {},
+   "source": [
+    "###  Appendix D5: Multiplying two different real numbers\n",
+    "Goal: Given quantum registers A and B with real numbers $\\kappa$ ($0 \\leq \\kappa \\lt 1$) and $\\lambda$ ($0 \\leq \\lambda \\lt 1$), our goal is to compute an approximation $\\widetilde{\\gamma}$ of $\\gamma = \\kappa * \\lambda$ s.t. $|\\widetilde{\\gamma} - \\gamma| < \\epsilon$\n",
+    "\n",
+    "$$\n",
+    "    |\\kappa\\rangle_{A} |\\lambda\\rangle_{B} |0\\rangle_{\\text{out}} \\rightarrow |\\kappa\\rangle_{A} |\\lambda\\rangle_{B} |\\widetilde{\\gamma}\\rangle_{\\text{out}}\n",
+    "$$\n",
+    "\n",
+    "\n",
+    "We analyze the error due to fixed point arithmetic for algorithm described in Appendix D5 of https://arxiv.org/abs/2007.07391"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "id": "b89122eb-d6b0-48f6-a3f4-4d3a0824ba7b",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Success! d=13, eps=0.001\n",
+      "Success! d=17, eps=0.0001\n",
+      "Success! d=21, eps=1e-05\n",
+      "Success! d=24, eps=1e-06\n",
+      "Success! d=28, eps=1e-07\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Multiplying two real numbers\n",
+    "def get_bitsize_for_fxp_mul(eps: float):\n",
+    "    return int(np.ceil(1 + np.log2(1/eps)) + np.log2(1 + np.log2(1/eps)))\n",
```

Comment:
this is from the paper?

Reply:
Yes, I'll add a comment to reference equation.